activerecord/lib/active_record/autosave_association.rb

```diff
@@ -447,22 +444,27 @@ def save_has_one_association(reflection)
           primary_key = Array(compute_primary_key(reflection, self)).map(&:to_s)
           primary_key_value = primary_key.map { |key| _read_attribute(key) }
 
-          if (autosave && record.changed_for_autosave?) || _record_changed?(reflection, record, primary_key_value)
```

Comment:
This looks like a lot of changes but all I've done is de-nest the logic inside this condition, and move the inverse `belongs_to` guard clause I originally added after it to so it can account for the record changed case.